include/deal.II/dofs/dof_tools.templates.h

```diff
@@ -0,0 +1,2882 @@
+    const auto &fe_collection = dof_handler.get_fe_collection();
+    AssertIndexRange(component, fe_collection.n_components());
+    for (unsigned int i = 0; i < fe_collection.size(); ++i)
+      {
+        Assert(fe_collection[i].is_primitive() == true,
+               typename FiniteElement<dim>::ExcFENotPrimitive());
+      }
+
+    // store a flag whether we should care about different components. this
+    // is just a simplification, we could ask for this at every single
+    // place equally well
+    const bool consider_components =
+      (dof_handler.get_fe_collection().n_components() != 1);
+
+    // zero out the components that we will touch
+    if (consider_components == false)
+      dof_data = 0;
+    else
+      {
+        std::vector<unsigned char> component_dofs(
+          dof_handler.n_locally_owned_dofs());
+        internal::get_component_association(
+          dof_handler,
+          fe_collection.component_mask(FEValuesExtractors::Scalar(component)),
+          component_dofs);
+
+        for (unsigned int i = 0; i < dof_data.size(); ++i)
+          if (component_dofs[i] == static_cast<unsigned char>(component))
+            dof_data(i) = 0;
+      }
+
+    // count how often we have added a value in the sum for each dof
+    std::vector<unsigned char> touch_count(dof_handler.n_dofs(), 0);
+
+    std::vector<types::global_dof_index> dof_indices;
+    dof_indices.reserve(fe_collection.max_dofs_per_cell());
+
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        const unsigned int dofs_per_cell = cell->get_fe().n_dofs_per_cell();
+        dof_indices.resize(dofs_per_cell);
+        cell->get_dof_indices(dof_indices);
+
+        for (unsigned int i = 0; i < dofs_per_cell; ++i)
+          // consider this dof only if it is the right component. if there
+          // is only one component, short cut the test
+          if (!consider_components ||
+              (cell->get_fe().system_to_component_index(i).first == component))
+            {
+              // sum up contribution of the present_cell to this dof
+              dof_data(dof_indices[i]) += cell_data(cell->active_cell_index());
+
+              // note that we added another summand
+              ++touch_count[dof_indices[i]];
+            }
+      }
+
+    // compute the mean value on all the dofs by dividing with the number
+    // of summands.
+    for (types::global_dof_index i = 0; i < dof_handler.n_dofs(); ++i)
+      {
+        // assert that each dof was used at least once. this needs not be
+        // the case if the vector has more than one component
+        Assert(consider_components || (touch_count[i] != 0),
+               ExcInternalError());
+        if (touch_count[i] != 0)
+          dof_data(i) /= touch_count[i];
+      }
+  }
+
+
+
+  template <int dim, int spacedim>
+  IndexSet
+  extract_dofs(const DoFHandler<dim, spacedim> &dof,
+               const ComponentMask             &component_mask)
```

Comment:
I don't see the point to have these sorts of functions in a `.templates.h` file. We know exactly what kinds of arguments these functions could possibly be instantiated for, and we do. There is no reason anyone could need these functions in a file that can be `#include`d.